prov/efa/src/efa_user_info.c

```diff
@@ -302,38 +303,58 @@ bool efa_user_info_should_support_hmem(int version, const struct fi_info *hints)
 	 */
 	if (FI_VERSION_GE(version, FI_VERSION(1, 18)) && hmem_ops[FI_HMEM_CUDA].initialized) {
 		EFA_INFO(FI_LOG_CORE,
-			"User is using API version >= 1.18. CUDA library and device are availble, claim support of FI_HMEM.");
+			"User is using API version >= 1.18. CUDA library and "
+			"device are availble, claim support of FI_HMEM.");
 		return true;
 	}
 
 	if (ofi_hmem_p2p_disabled()) {
 		EFA_WARN(FI_LOG_CORE,
 			"FI_HMEM capability currently requires peer to peer "
-			"support, which is disabled because FI_HMEM_P2P_DISABLED was set to 1/on/true.");
+			"support, which is disabled because "
+			"FI_HMEM_P2P_DISABLED was set to 1/on/true.");
 		return false;
 	}
 
-	if (!efa_device_support_rdma_read()) {
-		EFA_WARN(FI_LOG_CORE,
-			"FI_HMEM capability requires RDMA, which this device does not support.");
-		return false;
+	if (rxr_env_or_default_use_device_rdma()) {
+		/* support FI_HMEM because we can use RDMA. */
+		return true;
 	}
+	/* Otherwise, we cannot support FI_HMEM.  Log some relevant reason. */
 
-	if (!rxr_env_get_use_device_rdma()) {
+	if (!hmem_ops[FI_HMEM_CUDA].initialized) {
+		extra_info = "no CUDA library or device was found.";
+	}
+	else {
+		assert(!FI_VERSION_GE(version, FI_VERSION(1, 18)));
+		extra_info = "the application is using libfabric "
+			"API <1.18, so CUDA cannot be used.";
+	}
+
+	if (false == rxr_get_default_use_device_rdma()) {
 		EFA_WARN(FI_LOG_CORE,
-			"FI_HMEM capability requires RDMA, which is turned off. "
-			"You can turn it on by setting environment variable "
-			"FI_EFA_USE_DEVICE_RDMA to 1.\n");
-		return false;
+			"EFA cannot support FI_HMEM because the EFA device "
+			"does not have RDMA support and %s\n",
+			extra_info);
+	} else if (0 == rxr_env_get_use_device_rdma()) {
+		EFA_WARN(FI_LOG_CORE,
+			"EFA cannot support FI_HMEM because the environment "
+			"variable FI_EFA_USE_DEVICE_RDMA is 0 and %s\n"
+			"You can enable FI_HMEM by setting "
```

Comment:
I think we need to remove "You can xxx".

Reply:
because you think it's redundant or misleading?

Reply:
I think it is misleading because at this point there is a possiblity that CUDA device or CUDA library is not available. Therefore simply set the envar does not guarantee user will get FI_HMEM support.

Reply:
Removed, although at this point we know that CUDA device and library are available.